src/common/async/call_once.h

```diff
@@ -0,0 +1,189 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright contributors to the Ceph project
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation. See file COPYING.
+ *
+ */
+
+#pragma once
+
+#include <concepts>
+#include <condition_variable>
+#include <mutex>
+#include <variant>
+
+#include <boost/intrusive/list.hpp>
+
+#include "service.h"
+#include "yield_context.h"
+
+#include "detail/call_once.h"
+
+namespace ceph::async {
+
+/// Stores the result of call_once(), if any.
+///
+/// The cached Result type must be semiregular (copyable and default-
+/// constructible).
+template <std::semiregular Result>
+class once_flag final : public service_list_base_hook {
+ public:
+  once_flag() = default;
+  ~once_flag()
+  {
+    if (svc) {
+      svc->remove(*this);
+    }
+  }
+
+  /// Call the given function and cache its result. If a result is already
+  /// available, return it. If a call is already in progress, wait for its
+  /// completion and return its result. Otherwise, call f() and return its
+  /// result. f() must return Result when called with no arguments. If it
+  /// throws, that exception is rethrown by all calls.
+  template <typename Callable>
+  friend Result call_once(once_flag& self, optional_yield y, Callable&& f)
+      requires std::convertible_to<std::invoke_result_t<Callable>, Result>
+  {
+    auto lock = std::unique_lock{self.mutex};
+    return std::visit(CallVisitor{self, lock, y, std::forward<Callable>(f)},
```

Comment:
I'm going to have to skip ahead and read about the adventures of CallVisitor first! Looks good.